mailet/standard/src/main/java/org/apache/james/transport/mailets/StripAttachment.java

```diff
@@ -221,122 +223,136 @@ public void service(Mail mail) throws MailetException {
      * filename that matches the pattern. If the name matches, the content of
-     * the part is stored (using its name) in te given diretcory.
+     * the part is stored (using its name) in te given directory.
      * 
      * Note: this method is recursive.
      * 
      * @param part
      *            The part to analyse.
      * @param mail
-     * @return
+     * @return True if one of the subpart was removed
      * @throws Exception
      */
-    private boolean analyseMultipartPartMessage(Part part, Mail mail)
-            throws Exception {
-        if (part.isMimeType("multipart/*")) {
-            try {
-                Multipart multipart = (Multipart) part.getContent();
-                boolean atLeastOneRemoved = false;
-                int numParts = multipart.getCount();
-                for (int i = 0; i < numParts; i++) {
-                    Part p = multipart.getBodyPart(i);
-                    if (p.isMimeType("multipart/*")) {
-                        atLeastOneRemoved |= analyseMultipartPartMessage(p,
-                                mail);
-                    } else {
-                        boolean removed = checkMessageRemoved(p, mail);
-                        if (removed) {
-                            multipart.removeBodyPart(i);
-                            atLeastOneRemoved = true;
-                            i--;
-                            numParts--;
-                        }
+    @VisibleForTesting boolean analyseMultipartPartMessage(Part part, Mail mail) throws MailetException {
+        if (!isMultipart(part)) {
+            return false;
+        }
+
+        try {
+            Multipart multipart = (Multipart) part.getContent();
+            boolean atLeastOneRemoved = false;
+            List<BodyPart> bodyParts = retrieveBodyParts(multipart);
+            for (BodyPart bodyPart: bodyParts) {
+                if (isMultipart(bodyPart)) {
+                    atLeastOneRemoved |= analyseMultipartPartMessage(bodyPart, mail);
+                } else {
+                    if (processBodyPart(bodyPart, mail)) {
+                        multipart.removeBodyPart(bodyPart);
+                        atLeastOneRemoved = true;
                     }
                 }
-                if (atLeastOneRemoved) {
-                    part.setContent(multipart);
-                    if (part instanceof Message) {
-                        ((Message) part).saveChanges();
-                    }
+            }
+            if (atLeastOneRemoved) {
+                part.setContent(multipart);
+                if (part instanceof Message) {
+                    ((Message) part).saveChanges();
                 }
-                return atLeastOneRemoved;
-            } catch (Exception e) {
-                log("Could not analyse part.", e);
             }
+            return atLeastOneRemoved;
+        } catch (Exception e) {
+            log("Could not analyse part.", e);
```

Comment:
Message enhanced, the original exception is printed in the log so the admin will have the stacktrace.